lib/text/ssa_text_parser.js

```diff
@@ -0,0 +1,316 @@
+
+    const section = {
+      styles: '',
+      events: '',
+    };
+
+    const parts = str.split(/\r?\n\s*\r?\n/);
+    for (const part of parts) {
+      // SSA content
+      const match = SsaTextParser.ssaContent_.exec(part);
+      if (match) {
+        const tag = match[1];
+        const lines = match[2];
+        if (tag == 'V4 Styles' || tag == 'V4+ Styles') {
+          section.styles = lines;
+          continue;
+        }
+        if (tag == 'Events') {
+          section.events = lines;
+          continue;
+        }
+      }
+      shaka.log.warning('SsaTextParser parser encountered an unknown part.',
+          part);
+    }
+
+    // Process styles
+    const styles = [];
+
+    // Used to be able to iterate over the style parameters.
+    let styleColumns = null;
+
+    const styleLines = section.styles.split(/\r?\n/);
+    for (const line of styleLines) {
+      if (/^\s*;/.test(line)) {
+        // Skip comment
+        continue;
+      }
+      const lineParts = SsaTextParser.lineParts_.exec(line);
+      if (lineParts) {
+        const name = lineParts[1].trim();
+        const value = lineParts[2].trim();
+        if (name == 'Format') {
+          styleColumns = value.split(SsaTextParser.valuesFormat_);
+          continue;
+        }
+        if (name == 'Style') {
+          const values = value.split(SsaTextParser.valuesFormat_);
+          const style = {};
+          for (let c = 0; c < styleColumns.length && c < values.length; c++) {
+            style[styleColumns[c]] = values[c];
+          }
+          styles.push(style);
+          continue;
+        }
+      }
+    }
+
+    // Process cues
+    /** @type {!Array.<!shaka.extern.Cue>} */
+    const cues = [];
+
+    // Used to be able to iterate over the event parameters.
+    let eventColumns = null;
+
+    const eventLines = section.events.split(/\r?\n/);
+    for (const line of eventLines) {
+      if (/^\s*;/.test(line)) {
+        // Skip comment
+        continue;
+      }
+      const lineParts = SsaTextParser.lineParts_.exec(line);
+      if (lineParts) {
+        const name = lineParts[1].trim();
+        const value = lineParts[2].trim();
+        if (name == 'Format') {
+          eventColumns = value.split(SsaTextParser.valuesFormat_);
+          continue;
+        }
+        if (name == 'Dialogue') {
+          const values = value.split(SsaTextParser.valuesFormat_);
+          const data = {};
+          for (let c = 0; c < eventColumns.length && c < values.length; c++) {
+            data[eventColumns[c]] = values[c];
+          }
+
+          const startTime = SsaTextParser.parseTime_(data['Start']);
+          const endTime = SsaTextParser.parseTime_(data['End']);
+
+          // Note: Normally, you should take the "Text" field, but if it
+          // has a comma, it fails.
+          const indexOfPayload =
+            value.split(',', eventColumns.length - 1).join(',').length + 1;
+          const payload = value.substr(indexOfPayload)
```

Comment:
You could also just do `values.slice(eventColumns.length).join(',')` to get the payload.

Reply:
Changed!